modules/openapi-generator/src/main/resources/elixir/api.mustache

```diff
@@ -48,17 +48,18 @@ defmodule {{moduleName}}.Api.{{classname}} do
       :body => :body
 {{/isBodyParam}}
 {{^isBodyParam}}
-      {{#atom}}{{baseName}}{{/atom}} => {{#isFormParam}}:form{{/isFormParam}}{{#isQueryParam}}:query{{/isQueryParam}}{{#isHeaderParam}}:headers{{/isHeaderParam}}{{^-last}},{{/-last}}
+      {{#atom}}{{&baseName}}{{/atom}} => {{#isFormParam}}:form{{/isFormParam}}{{#isQueryParam}}:query{{/isQueryParam}}{{#isHeaderParam}}:headers{{/isHeaderParam}}{{^-last}},{{/-last}}
 {{/isBodyParam}}
 {{#-last}}
     }
 
 {{/-last}}
 {{/optionalParams}}
     request =
-      %{}
-      |> method({{#atom}}{{#underscored}}{{httpMethod}}{{/underscored}}{{/atom}})
-      |> url("{{replacedPathName}}")
+      %Tesla.Env{
+        method: {{#atom}}{{#underscored}}{{httpMethod}}{{/underscored}}{{/atom}},
+        url: "{{replacedPathName}}",
```

Comment:
I’m not sure why this change was made. Yes, it saves two function calls, but it also means that it is harder to extract Tesla from the execution path in favour of https://github.com/wojtekmach/req.

(A different set of changes that I’m going to be looking at will be to *replace* the structs with validated maps, because there’s a *lot* of atoms generated in a sufficiently complex API.)

Reply:
I can revert this one. It helped me with debugging the generated code. 

It shows from the get-go that the pipe inputs/outputs are a `Tesla.Env`.
I think a good way forward with `Tesla` would be to just compose a middleware stack instead of piping, which cannot be altered at runtime.

I don't have experience with [req](https://github.com/wojtekmach/req.), but I can look into it if you are interested in also having continued support for Tesla.  

> (A different set of changes that I’m going to be looking at will be to _replace_ the structs with validated maps, because there’s a _lot_ of atoms generated in a sufficiently complex API.)

Do you want to completely do away with structs?
I concur with you on not creating atoms to the extent that they should not be generated _from input_ (e.g. for values supplied in function calls).

Here, on the other hand, we do have a finite number of atoms, as each struct generates only it's fields' descriptors as atoms at compile time.

So IMHO this is not a case of unbounded atom generation. Ex/Beam is good at handling lots of atoms, their number should just not be unbounded or dependent on runtime and unsafe input.

### Utility module for assemling middelwares ? 
If you want I can look at creating a utility Module to handle a list of middlewares instead of the current approach. The Api functions would then only use this utility class to insert the necessary middlewares to set the url, params e.t.c.

This way the utility Module would allow for runtime transformations, even outside the spec (e.g. caching, delegation to other nodes e.t.c. from client ).   

Do you want to do away completely with structs ?

Reply:
I'm not quite sure about this discussion. 

Since it's not related to the implementation of `AnyType`, it would be great to bring this concern into an issue. This would also allow the discussion a more retractable character.

For the change itself: (by itself, I just mean the line without the changing-possibility and debugability it could bring)

I can't see the additional value it brings for breaking the style of composing/piping, and therefore I can't just let it through.

I would appreciate it if you could extract this change into a different branch and start the discussion with an issue. You could also link the new branch in the form of a PR, so it doesn't get lost.

Reply:
Ok, will come back with a new pr, on